File: go-ms/infrastructure/driven-adapters/db/account/account_data_repository.go
package data

import (
	"fmt"
	model "maocq/go-ms/domain/model/account"
	data "maocq/go-ms/infrastructure/driven-adapters/db/account/data"

	"gorm.io/gorm"
)

type AccountDataRepository struct {
	DB *gorm.DB
}

func (r *AccountDataRepository) FindById(id int) *model.Account {
	var accountData data.AccountData
	if result := r.DB.First(&accountData, id); result.Error != nil {
		fmt.Println("Error", result.Error)
	}

	return toEntity(&accountData)
}

func (r *AccountDataRepository) FindOne() []*model.Account {
	var accountsData []data.AccountData
	if result := r.DB.Where("id = ?", 4000).Find(&accountsData); result.Error != nil {
		fmt.Println("Error", result.Error)
	}
	var accounts []*model.Account
	for _, account := range accountsData {
		accounts = append(accounts, toEntity(&account))
	}

	return accounts
}

func (r *AccountDataRepository) Update(account *model.Account) *model.Account {
	data := toData(account)
	if result := r.DB.Save(data); result.Error != nil {
		fmt.Println("Error", result.Error)
	}
	return toEntity(data)
}

func toEntity(data *data.AccountData) *model.Account {
	return &model.Account{Id: data.Id, UserId: data.UserId, Account: data.Account, Name: data.Name,
		Number: data.Number, Balance: data.Balance, Currency: data.Currency, Type: data.Type,
		Bank: data.Bank, CreationDate: data.CreationDate, UpdateDate: data.UpdateDate}
}

func toData(model *model.Account) *data.AccountData {
	return &data.AccountData{Id: model.Id, UserId: model.UserId, Account: model.Account,
		Name: model.Name, Number: model.Number, Balance: model.Balance, Currency: model.Currency,
		Type: model.Type, Bank: model.Bank, CreationDate: model.CreationDate, UpdateDate: model.UpdateDate}
}

File: elixir-ms/lib/infraestructure/driven-adapters/db/repository/account/account_data_repository.ex
defmodule ElixirMs.DrivenAdapters.Db.Repository.AccountDataRepository do
  alias ElixirMs.Repo
  alias ElixirMs.DrivenAdapters.Db.Repository.AccountData
  alias ElixirMs.Model.Account

  @behaviour ElixirMs.Model.Behaviour.AccountRepository

  def find_by_id(id), do: AccountData |> Repo.get!(id) |> to_entity
  def find_one(), do: find_one_sql().rows |> Enum.map(&row_to_entity/1)
  def find_test(), do: find_test_sql().rows |> Enum.map(&row_to_entity/1)
  def find_all(), do: AccountData |> Repo.all |> Enum.map(&to_entity/1)
  def update(entity) do
    row = %AccountData{id: entity.id}
    Repo.update!(Ecto.Changeset.change(row, %{user_id: entity.user_id, account: entity.account, name: entity.name, number: entity.number,
     balance: entity.balance, currency: entity.currency, type: entity.type, bank: entity.bank,
     creation_date: entity.creation_date, update_date: entity.update_date})) |> to_entity
  end

  defp find_one_sql() do
    query = "SELECT a.id, a.user_id, a.account, a.name, a.number, a.balance, a.currency, a.type, a.bank, a.creation_date, a.update_date FROM account AS a WHERE (a.id = 4000)"
    Ecto.Adapters.SQL.query!(Repo, query, [])
  end

  defp find_test_sql() do
    query = "select * from account a where a.balance > (select avg(a.balance) + 72700000 from account a)"
    Ecto.Adapters.SQL.query!(Repo, query, [])
  end


  defp to_entity(data) do
    %Account{id: data.id, user_id: data.user_id, account: data.account, name: data.name, number: data.number,
     balance: data.balance, currency: data.currency, type: data.type, bank: data.bank,
     creation_date: data.creation_date, update_date: data.update_date}
  end

  defp row_to_entity(data) do
    %Account{id: data |> Enum.at(0), user_id: data |> Enum.at(1), account: data |> Enum.at(2), name: data |> Enum.at(3), number: data |> Enum.at(4),
     balance: data |> Enum.at(5), currency: data |> Enum.at(6), type: data |> Enum.at(7), bank: data |> Enum.at(8),
     creation_date: data |> Enum.at(9), update_date: data |> Enum.at(10)}
  end

  defp to_row(entity) do
    row = %AccountData{id: entity.id, user_id: entity.user_id, account: entity.account, name: entity.name, number: entity.number,
     balance: entity.balance, currency: entity.currency, type: entity.type, bank: entity.bank,
     creation_date: entity.creation_date, update_date: entity.update_date}
  end
end
